### PR TITLE
IMGW: calculate distances on spheroid (fix #60)

### LINKED ISSUE #60: Calculations in WGS 84 are inaccurate
I think you should consider saving the location of IMGW stations in EPSG 2180 coordinate system as default. Currently distance calculations based on WGS 84 are inaccurate (several km difference).

```r
library("sf")
library("climate")

data(imgw_hydro_stations)
imgw_hydro_stations = imgw_hydro_stations[imgw_hydro_stations$id == 154170050, ]
coords = c(imgw_hydro_stations$X, imgw_hydro_stations$Y)
st = nearest_stations_imgw(type = "meteo", rank = "synop", point = coords,
                           no_of_stations = 1, add_map = FALSE)
st$`distance [km]`
#> 35.601

### EPSG 2180
P1 = st_sfc(st_point(coords), crs = 4326)
P1 = st_transform(P1, crs = 2180)
P2 = st_sfc(st_point(c(st$X, st$Y)), crs = 4326)
P2 = st_transform(P2, crs = 2180)

as.vector(st_distance(P1, P2)) / 1000
#> 21.01992
```

### PR DESCRIPTION
**Pull request by kadyb**
I added `spheroid_dist()` function that calculates distances between points on a spheroid using Vincenty's formula. This is currently only used by `nearest_stations_imgw()`. Additionally, in this PR I improved the code style. Two other methods of calculating distances are mentioned [here](https://github.com/bczernecki/climate/issues/60#issuecomment-987176584). 

```r
library("climate")

# select GARDNA WIELKA
imgw_hydro_stations = imgw_hydro_stations[imgw_hydro_stations$id == 154170050, ]
coords = c(imgw_hydro_stations$X, imgw_hydro_stations$Y)
st = nearest_stations_imgw(type = "meteo", rank = "synop", point = coords,
                           no_of_stations = 1, add_map = FALSE)

st[, 5]

# BEFORE PR: 35.601 km
# AFTER PR: 20.964 km
# EPSG 2180: 21.01992 km
```

### CODE REVIEW

File: R/nearest_stations_imgw.R [hunk 1]
```diff
@@ -105,10 +105,10 @@ nearest_stations_imgw = function(type = "meteo",
       pch = 19,
       xlab = "longitude",
       ylab = "latitude",
-      xlim = (c(min(c(result$X, point$X), na.rm = TRUE) - 1,
-                max(c(result$X, point$X), na.rm = TRUE) + 1)),
-      ylim = (c(min(c(result$Y, point$Y), na.rm = TRUE) - 1,
-                max(c(result$Y, point$Y), na.rm = TRUE) + 1)),
+      xlim = c(min(c(result$X, point[1]), na.rm = TRUE) - 1,
```
Comment (bczernecki) [inline R/nearest_stations_imgw.R:108]: mixing dolars with square brackets is not a good practice

File: R/nearest_stations_imgw.R [hunk 2]
```diff
@@ -25,17 +25,17 @@
...
+                                 add_map = TRUE,
+                                 point = NULL,
+                                 no_of_stations = 50,
+                                 ...) {
   if (length(point) > 2) {
     stop("Too many points for the distance calculations. Please provide just one pair of coordinates (e.g. point = c(17,53))")
   } else if (length(point) < 2) {
-    message("
-            The point argument should have two coordinates.
+    message("The point argument should have two coordinates.
```
Comment (bczernecki) [inline R/nearest_stations_imgw.R:38]: This will make a message indention which look quite strange. I'd rather stick to the previous version even if the code looks not that nice

File: R/nearest_stations_imgw.R [hunk 3]
```diff
@@ -25,17 +25,17 @@
...
+                                 rank = "synop",
+                                 year = 2018,
+                                 add_map = TRUE,
+                                 point = NULL,
+                                 no_of_stations = 50,
+                                 ...) {
   if (length(point) > 2) {
     stop("Too many points for the distance calculations. Please provide just one pair of coordinates (e.g. point = c(17,53))")
   } else if (length(point) < 2) {
-    message("
```
Comment (bczernecki) [inline R/nearest_stations_imgw.R:37]: message with indentions look bad

File: R/nearest_stations_imgw.R [hunk 4]
```diff
@@ -60,7 +60,7 @@ nearest_stations_imgw = function(type = "meteo",
     stop("You've provided wrong type argument; please use: \"meteo\", or \"hydro\"")
   }
 
-  if (dim(result)[1] == 0) {
+  if (nrow(result) == 0) {
```
Comment (bczernecki) [inline R/nearest_stations_imgw.R:63]: what's the rationale behind this change?

**Comment by bczernecki** [comment]
First of all, thank you @kadyb for the work on fixing this bug! We do appreciate it!
Please fix all changes according to provided comments and create new PR pointing to devel branch.
Thanks again!

File: NEWS.md [hunk 5]
```diff
@@ -1,103 +1,108 @@
...
-    * New function `hydro_daily()` for downloading daily hydrological data
-    * New function `meteo_metadata()` for downloading the metadata of the meteorological data
-    * New function `hydro_metadata()` for downloading the metadata of the hydrological data
-    * New function `meteo_sounding()` for downloading the mea (i.e. measurements of the vertical profile of atmosphere) sounding data
-* New datasets
-    * New dataset `meteo_stations` containing Polish meteorological station's localizations
-    * New dataset `hydro_stations` containing Polish hydrological station's localizations
-* Improvements
-    * Added a `NEWS.md` file to track changes to the package.
+# climate 1.0.4
```
Comment (bczernecki) [inline NEWS.md:1]: There's only one change - not > 1100

File: R/nearest_stations_imgw.R [hunk 6]
```diff
@@ -5,83 +5,92 @@
...
+    if (any(colnames(result) == "LON"))
+      point = c(round(mean(result$LON, na.rm = TRUE), 2),
+                round(mean(result$LAT, na.rm = TRUE), 2))
+    if (any(colnames(result) == "X"))
+      point = c(round(mean(result$X, na.rm = TRUE), 2),
+                round(mean(result$Y, na.rm = TRUE), 2))
   }
   
+  dist_vec = double(nrow(result))
+  for (i in seq_along(dist_vec)) {
```
Comment (bczernecki) [inline R/nearest_stations_imgw.R:79]: please vectorise if possible (if not - leave as is)

**Review by bczernecki — CHANGES_REQUESTED** [review]

File: R/nearest_stations_imgw.R (hunk 4 above)
Comment (kadyb) [reply R/nearest_stations_imgw.R:63]: `result` is a simple data.frame so in my opinion it is better to make it clear (for readability) that we are using the number of rows. `dim()` can refer to multidimensional arrays but behind the code `nrow()` is just `dim()[1]`.

File: NEWS.md (hunk 5 above)
Comment (kadyb) [reply NEWS.md:1]: Yeah, probably this is a problem with file encoding between Windows and Linux. I will revert it and edit this file directly on GitHub.

File: R/nearest_stations_imgw.R (hunk 6 above)
Comment (kadyb) [reply R/nearest_stations_imgw.R:79]: OK, I will vectorise this, but please note that vector preallocation and value assignment is faster and requires less memory than vectorization. On the other hand, there are probably not too many locations (thousands?), so it shouldn't be a problem.

**Comment by kadyb** [comment]
My questions:

1. What is the reason to wait 2 seconds if the number of coordinates is less than 2? 
https://github.com/bczernecki/climate/blob/f96151442a9a0806b2dfa44c3811adc33449a879/R/nearest_stations_imgw.R#L41

2. Possibly this line can be deleted?
https://github.com/bczernecki/climate/blob/f96151442a9a0806b2dfa44c3811adc33449a879/R/nearest_stations_imgw.R#L13

File: R/nearest_stations_imgw.R (hunk 2 above)
Comment (kadyb) [reply R/nearest_stations_imgw.R:38]: What do you think about this?

```r
text = paste("The point argument should have two coordinates.",
             "We will provide nearest stations for mean location of all available stations.",
             "To change it please change the `point` argument c(LON,LAT)",
             sep = "\n")
message(text)
```

File: R/nearest_stations_imgw.R (hunk 4 above)
Comment (bczernecki) [reply R/nearest_stations_imgw.R:63]: I think both approaches are fine. I don't remember precisely, but maybe we wanted to have a passing code for arrays that might be converted to vector and `dim` would work for any class of object, but I am only guessing...

File: R/nearest_stations_imgw.R (hunk 6 above)
Comment (bczernecki) [reply R/nearest_stations_imgw.R:79]: I agree, that can be left as is

**Comment by bczernecki** [comment]
>     1. What is the reason to wait 2 seconds if the number of coordinates is less than 2?
>        https://github.com/bczernecki/climate/blob/f96151442a9a0806b2dfa44c3811adc33449a879/R/nearest_stations_imgw.R#L41
Message is shown and user can read it. Otherwise further chunks of code are executed and in order to read the message user needs to scroll up.
>
>     2. Possibly this line can be deleted?
>        https://github.com/bczernecki/climate/blob/f96151442a9a0806b2dfa44c3811adc33449a879/R/nearest_stations_imgw.R#L13
Agree

File: R/nearest_stations_imgw.R (hunk 4 above)
Comment (kadyb) [reply R/nearest_stations_imgw.R:63]: OK, let's stay with `dim(result)[1]` to avoid changing too much. BTW: Using `dim()` with objects of another class (not matrices / data frames) can produce results that are completely unexpected by us. For example, stars object from [stars](https://r-spatial.github.io/stars/) package. It uses arrays on backend but first dimension are attributes, second are columns, third are rows, four are bands, etc.

File: R/nearest_stations_imgw.R (hunk 6 above)
Comment (kadyb) [reply R/nearest_stations_imgw.R:79]: Just in case, reference for future:

```r
dist_vec = apply(X = result[, 2:3], MARGIN = 1, FUN = spheroid_dist, p1 = point)
```

**Comment by kadyb** [comment]
> Message is shown and user can read it. Otherwise further chunks of code are executed and in order to read the message user needs to scroll up.

I wouldn't think of it! So I will add a comment with an explanation and I add `Sys.sleep(2)` to this too:
https://github.com/bczernecki/climate/blob/f96151442a9a0806b2dfa44c3811adc33449a879/R/nearest_stations_imgw.R#L45-L48

**Review by bczernecki — APPROVED** [review]
LGTM 👍

**Comment by kadyb** [comment]
Thanks for the comments and acceptance this PR! 

There is an error in `spheroid_dist.R`:
```
> ## No test: 
>  p1 = c(18.633333, 54.366667) # Gdańsk
>  p2 = c(17.016667, 54.466667) # Słupsk
>  spheroid_dist(p1, p2)
Error: Error in spheroid_dist(p1, p2) : could not find function "spheroid_dist"
```

We should:
1. change `\donttest` to `\dontrun`
2. or use `climate:::spheroid_dist(p1, p2)` instead of `spheroid_dist(p1, p2)`

**Comment by bczernecki** [comment]
Yes, I have added `spheroid_dist` to **NAMESPACE**, improved documentation and also added test to increase package coverage. It is now committed on the dev branch.
Thanks again for your contribution to the climate package!

**Comment by kadyb** [comment]
I thought `spheroid_dist()` shouldn't be available to users (so I didn't export it), but as global function it is fine too.
